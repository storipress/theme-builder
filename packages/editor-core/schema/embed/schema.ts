export { EmbedSchema } from '@storipress/tiptap-schema'
