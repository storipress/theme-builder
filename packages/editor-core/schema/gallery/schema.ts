export { GallerySchema } from '@storipress/tiptap-schema'
