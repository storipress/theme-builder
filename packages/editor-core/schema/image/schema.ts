export { ImageSchema } from '@storipress/tiptap-schema'
