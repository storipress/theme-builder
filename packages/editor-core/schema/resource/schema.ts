export { ResourceSchema } from '@storipress/tiptap-schema'
