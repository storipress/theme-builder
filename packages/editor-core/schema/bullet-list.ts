export { BulletList } from '@storipress/tiptap-schema'
