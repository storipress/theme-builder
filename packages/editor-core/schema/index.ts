export { baseExtensions as schemaExtensions } from './extensions'
export { warnExceedSizeLimit } from './schema-bus'
export { render, schema } from './schema-only'
