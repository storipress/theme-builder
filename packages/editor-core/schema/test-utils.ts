import { Document } from '@tiptap/extension-document'
import { Paragraph } from '@tiptap/extension-paragraph'
import { Text } from '@tiptap/extension-text'

export const baseSchema = [Document, Paragraph, Text]
