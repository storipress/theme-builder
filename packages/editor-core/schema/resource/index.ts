import type { Node } from '@tiptap/core'
import { VueNodeViewRenderer } from '@tiptap/vue-2'

import { stopEvent } from '../draggable-node'
import { nodePasteRule } from './node-paste-rule'
import { ResourceSchema } from './schema'
import ResourceView from './view.vue'

const URL_REGEX = /https?:\/\/(?:www\.)?[\w#%+.:=@~-]{1,256}\.[a-z]{2,}\b[\w#%&+,./:=?@~-]*/gi

const EMBED_REGEXES = [
  /youtube\.com\/watch/,
  /twitter\.com\/[^/]+\/status/,
  /instagram\.com\/p/,
  /vimeo\.com\/\d+/,
  /codepen\.io\/[^/]+\/pen/,
  // eslint-disable-next-line security/detect-unsafe-regex
  /soundcloud\.com(?:\/[^/]+){2}/,
]

export { ResourceSchema }

export function attachResourceNodeView(node: Node<unknown>): Node<unknown> {
  return node.extend({
    addPasteRules() {
      const { type } = this
      return [
        nodePasteRule(URL_REGEX, type, (match: string) => {
          if (EMBED_REGEXES.some((regex) => regex.test(match))) {
            return {
              type: 'embed',
              url: match,
            }
          }
          return {
            type: 'bookmark',
            url: match,
          }
        }),
      ]
    },
    addNodeView() {
      return VueNodeViewRenderer(ResourceView, { stopEvent })
    },
  })
}

export const Resource = attachResourceNodeView(ResourceSchema)
