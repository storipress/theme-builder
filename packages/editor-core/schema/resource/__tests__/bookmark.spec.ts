import type { BookmarkMeta } from '../../../core/api'
import { render } from '@testing-library/vue'
import { getHTMLFromFragment, getSchema } from '@tiptap/core'
import { Fragment } from 'prosemirror-model'
import { expect, it } from 'vitest'
import { baseSchema } from '../../test-utils'
import Bookmark from '../bookmark.vue'
import { ResourceSchema } from '../schema'
import '@testing-library/jest-dom/vitest'

const extensions = [...baseSchema, ResourceSchema]

it('have same html structure between bookmark.vue and render-resource-impl.jsx', () => {
  const schema = getSchema(extensions)
  const meta: BookmarkMeta = {
    url: '#',
    title: 'title',
    thumbnail: 'thumbnail',
    publisher: 'publisher',
    icon: 'icon',
    description: 'description',
    author: 'Author',
    aspectRadio: 1,
    html: '<div>title</div>',
  }
  const { getByText, container } = render(Bookmark, { props: { meta } })

  const title = getByText('title')

  expect(title).toBeInTheDocument()

  const staticHTML = getHTMLFromFragment(
    Fragment.fromArray([schema.nodes.resource.create({ meta: JSON.stringify(meta), type: 'bookmark' })]),
    schema
  )
  expect(staticHTML).toMatchSnapshot()
  expect(staticHTML).not.toContain('template')

  const div = document.createElement('div')
  div.innerHTML = staticHTML
  const bookmarkHTML = div.querySelector('.bookmark__link')?.innerHTML
  expect(bookmarkHTML).toBeDefined()
  expect(bookmarkHTML).toMatchSnapshot()
  expect(container.innerHTML).toMatchSnapshot()
})

it('render success when no author', () => {
  const schema = getSchema(extensions)
  const meta: BookmarkMeta = {
    url: '#',
    title: 'title',
    thumbnail: 'thumbnail',
    publisher: 'publisher',
    icon: 'icon',
    html: '<div>foo</div>',
    description: 'description',
    author: null,
    aspectRadio: 1,
  }
  expect(() => {
    const staticHTML = getHTMLFromFragment(
      Fragment.fromArray([schema.nodes.resource.create({ meta: JSON.stringify(meta), type: 'bookmark' })]),
      schema
    )

    expect(staticHTML).toMatchSnapshot()
  }).not.toThrow()
})

it('render success when no html', () => {
  const schema = getSchema(extensions)
  const meta: BookmarkMeta = {
    url: '#',
    title: 'title',
    thumbnail: 'thumbnail',
    publisher: 'publisher',
    icon: 'icon',
    html: '',
    description: 'description',
    author: null,
    aspectRadio: 1,
  }
  expect(() => {
    const staticHTML = getHTMLFromFragment(
      Fragment.fromArray([schema.nodes.resource.create({ meta: JSON.stringify(meta), type: 'bookmark' })]),
      schema
    )

    expect(staticHTML).toMatchSnapshot()
  }).not.toThrow()
})

it('render success for a lot of empty data', () => {
  const schema = getSchema(extensions)
  const meta: BookmarkMeta = {
    title: "Chairman's Letter - 1980",
    description: '',
    author: '',
    icon: '',
    url: 'https://www.berkshirehathaway.com/letters/1980.html',
    thumbnail: '',
    aspectRadio: 1,
  } as any
  expect(() => {
    const staticHTML = getHTMLFromFragment(
      Fragment.fromArray([schema.nodes.resource.create({ meta: JSON.stringify(meta), type: 'bookmark' })]),
      schema
    )

    expect(staticHTML).toMatchSnapshot()
  }).not.toThrow()
})
