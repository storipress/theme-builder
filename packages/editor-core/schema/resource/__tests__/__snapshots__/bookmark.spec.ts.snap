// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`have same html structure between bookmark.vue and render-resource-impl.jsx 1`] = `"<div class="clear-both opacity-100 relative" data-format="resource" data-meta="{&quot;url&quot;:&quot;#&quot;,&quot;title&quot;:&quot;title&quot;,&quot;thumbnail&quot;:&quot;thumbnail&quot;,&quot;publisher&quot;:&quot;publisher&quot;,&quot;icon&quot;:&quot;icon&quot;,&quot;description&quot;:&quot;description&quot;,&quot;author&quot;:&quot;Author&quot;,&quot;aspectRadio&quot;:1,&quot;html&quot;:&quot;<div>title</div>&quot;}" data-type="bookmark"><a class="bookmark__link"><div class="absolute inset-0 z-10"></div><figure><div class="bookmark__link"><div><div>title</div></div></div><figcaption class="text-center"></figcaption></figure></a></div>"`;

exports[`have same html structure between bookmark.vue and render-resource-impl.jsx 2`] = `"<div class="absolute inset-0 z-10"></div><figure><div class="bookmark__link"><div><div>title</div></div></div><figcaption class="text-center"></figcaption></figure>"`;

exports[`have same html structure between bookmark.vue and render-resource-impl.jsx 3`] = `"<div class="relative w-full"><div><div>title</div></div><div class="absolute inset-0 h-full w-full"></div></div>"`;

exports[`render success for a lot of empty data 1`] = `"<div class="clear-both opacity-100 relative" data-format="resource" data-meta="{&quot;title&quot;:&quot;Chairman's Letter - 1980&quot;,&quot;description&quot;:&quot;&quot;,&quot;author&quot;:&quot;&quot;,&quot;icon&quot;:&quot;&quot;,&quot;url&quot;:&quot;https://www.berkshirehathaway.com/letters/1980.html&quot;,&quot;thumbnail&quot;:&quot;&quot;,&quot;aspectRadio&quot;:1}" data-type="bookmark"><a class="bookmark__link"><div class="absolute inset-0 z-10"></div><figure><a class="bookmark__link" href="https://www.berkshirehathaway.com/letters/1980.html"><div class="bookmark"><div class="bookmark__info"><div class="bookmark__headline">Chairman's Letter - 1980</div><div class="bookmark__description"><div class="bookmark__content"></div></div><div class="bookmark__meta"><span></span></div></div></div></a><figcaption class="text-center"></figcaption></figure></a></div>"`;

exports[`render success when no author 1`] = `"<div class="clear-both opacity-100 relative" data-format="resource" data-meta="{&quot;url&quot;:&quot;#&quot;,&quot;title&quot;:&quot;title&quot;,&quot;thumbnail&quot;:&quot;thumbnail&quot;,&quot;publisher&quot;:&quot;publisher&quot;,&quot;icon&quot;:&quot;icon&quot;,&quot;html&quot;:&quot;<div>foo</div>&quot;,&quot;description&quot;:&quot;description&quot;,&quot;author&quot;:null,&quot;aspectRadio&quot;:1}" data-type="bookmark"><a class="bookmark__link"><div class="absolute inset-0 z-10"></div><figure><div class="bookmark__link"><div><div>foo</div></div></div><figcaption class="text-center"></figcaption></figure></a></div>"`;

exports[`render success when no html 1`] = `"<div class="clear-both opacity-100 relative" data-format="resource" data-meta="{&quot;url&quot;:&quot;#&quot;,&quot;title&quot;:&quot;title&quot;,&quot;thumbnail&quot;:&quot;thumbnail&quot;,&quot;publisher&quot;:&quot;publisher&quot;,&quot;icon&quot;:&quot;icon&quot;,&quot;html&quot;:&quot;&quot;,&quot;description&quot;:&quot;description&quot;,&quot;author&quot;:null,&quot;aspectRadio&quot;:1}" data-type="bookmark"><a class="bookmark__link"><div class="absolute inset-0 z-10"></div><figure><div class="bookmark__link"></div><figcaption class="text-center"></figcaption></figure></a></div>"`;
